docs: dynamic host volume command line (#24814)
hashicorp/nomad: website/content/docs/commands/volume/register.mdx

layout: docs
page_title: 'Commands: volume register'
description: |

>   Register CSI volumes or dynamic host volumes.

Reviewer: ```suggestion
  The `nomad volume register` command registers storage volumes as either Container Storage Interface (CSI) volumes or dynamic host volumes. The volume must already exist on a node or remote storage provider before you can register the volume for use in job task.
```

---

# Command: volume register

The `volume register` command registers storage volumes as either [Container
Storage Interface (CSI)][csi] volumes or dynamic host volumes. The volume must

> already exist on a node or remote storage provider before it can be registered
> and used by a task.

Reviewer: ```suggestion
already exist on a node or remote storage provider before you can register
the volume for use in job task.
```

To instead create a volume that does not already exist, use the [`volume
create`][] command. The `volume create` command automatically registers the
volume as well.

## Usage

```plaintext
nomad volume register [options] [file]
```

The `volume register` command requires a single argument, specifying the path to a
file containing a valid [volume specification][volume_specification]. Nomad
reads the file and submits the volume to the server for placement. If the

> supplied path is "-", the volume file is read from STDIN. Otherwise the file is
> read from the file at the supplied path.

Reviewer: ```suggestion
supplied path is "-", Nomad reads the volume file from STDIN. Otherwise Nomad reads the file from the supplied path.
```

When ACLs are enabled, this command requires a token with the appropriate
capability in the volume's namespace: the `csi-write-volume` capability for CSI
volumes or `host-volume-register` for dynamic host volumes.

## General Options

@include 'general_options.mdx'

## Volume Register Options

- `-policy-override`: Sets the flag to force override any soft mandatory
  Sentinel policies. Used for dynamic host volumes only. Not valid for CSI
  volumes.

## Volume Specification

The volume specification for CSI is documented in the [CSI Volume
Specification][csi_vol_spec] page. The volume specification for dynamic host
volumes is documented in the [Dynamic Host Volume Specification](host_vol_spec)
page.

> The `volume register` command allows updating a volume definition. But not all
> fields can be updated after the volume is registered.

Reviewer: ```suggestion
The `volume register` command allows updating a volume definition. However, you can only update some
fields after the volume is registered.
```

  volumes.

## Volume Specification

Refer to the [CSI Volume Specification][csi_vol_spec] and the [Dynamic Host
Volume Specification](host_vol_spec) for further information.

The `volume register` command allows updating a volume definition. However, you
can only update some fields after the volume is registered.

For CSI volumes:

- Update the `capacity_min` and `capacity_max` fields and increase
  the volume size if the CSI plugin supports it. Expansion may or may not be
  possible while the volume is in use, again depending on the plugin.
  Reducing volume capacity is not allowed per the CSI spec.
- Add or remove the `capability` blocks, but only if the capability is
  not currently in use by a mounted volume.
- Update the `mount_options` block if the volume is not in use.
- Update the `secrets` block.
- Update the `context` block. The values for this field are typically provided
  by the CSI plugin, and you should not update them unless recommended by the
  CSI plugin's documentation.

For dynamic host volumes:

- Update the `capacity_min` and `capacity_max` fields and increase
  the volume size if the plugin supports it. Expansion may or may not be

>   possible while the volume is in use, again depending on the plugin.

Author: Good catch. This code path doesn't change the size at all. Even `capacity` only records a changed size.

  Reducing volume capacity is not allowed.
- Update the `capacity` field only for volume registration.
- Update the `constraint` fields on `volume create`, but not after you create the volume.
- Add or remove the `capability` blocks but only if a mounted volume is not
  using currently using the capability.
- Change the `node_pool` field only from empty to the node pool that matches the

>   `node_id` field.

Author: The pool needs to match the node _if set_, so that's where this comes from. If you omit the pool on update, the server will fill it in correctly for you from the existing volume (`CanonicalizeForRegister`)
